odoo/addons/base/models/qweb.py

```diff
@@ -840,822 +628,545 @@ def _compile_static_node(self, el, options):
+                    {has_value} = True
+            """).strip(), indent))
+
+        code.append(self._indent(dedent(f"""
+                for index, item in enumerate({t_foreach}):
+                    values["{varname}_index"] = index
+                    if {has_value}:
+                        values["{varname}"], values["{varname}_value"] = item
+                    else:
+                        values["{varname}"] = values["{varname}_value"] = item
+                    values["{varname}_first"] = values["{varname}_index"] == 0
+                    if {size} is not None:
+                        values["{varname}_last"] = index + 1 == {size}
+                    values["{varname}_odd"] = index % 2
+                    values["{varname}_even"] = not values["{varname}_odd"]
+                    values["{varname}_parity"] = 'odd' if values["{varname}_odd"] else 'even'
+            """), indent))
+
+        if stop_profiling:
+            code.append(self._indent(f"""if values["{varname}_index"] == 0:""", indent + 1))
+            for line in stop_profiling:
+                code.append(self._indent(line, indent + 2))
+
+        code.append(self._indent(f"""log["last_path_node"] = "{self._compile_str(options['root'].getpath(el))}" """, indent + 1))
+        code.extend(content_foreach or self._indent('continue', indent + 1))
+
+        if stop_profiling:
+            code.append(self._indent(f"""if not {t_foreach}:""", indent))
+            for line in stop_profiling:
+                code.append(self._indent(line, indent + 1))
+
+        return code
+
+    def _compile_directive_esc(self, el, options, indent, name="esc"):
+        expr = el.attrib.pop('t-' + name)
+        directive = f't-{name}="{expr}"'
+        code = self._flushText(options, indent)
+        code_options = self._compile_options(el, f't_{name}_t_options', options, indent)
+        code.extend(code_options)
+        code.extend(self._compile_hook_before_directive(el, directive, options, indent))
 
-    def _compile_directive_groups(self, el, options):
-        return [
-            # if self.user_has_groups($groups):
-            #    next tag directive
-            ast.If(
-                test=ast.Call(
-                    func=ast.Attribute(
-                        value=ast.Name(id='self', ctx=ast.Load()),
-                        attr='user_has_groups',
-                        ctx=ast.Load()
-                    ),
-                    args=[ast.Str(el.attrib.pop('t-groups'))], keywords=[],
-                    starargs=None, kwargs=None
-                ),
-                body=self._compile_directives(el, options) or [ast.Pass()],
-                orelse=[]
-            )
-        ]
+        if expr == "0":
+            if code_options or name == "esc":
+                code.append(self._indent("""content = u''.join(values.get("0", []))""", indent))
+            else:
+                code.append(self._indent("""content = values.get("0", [])""", indent))
```

Comment:
In order not to make an unnecessary join, moreover in another branch (on the caches) it is necessary to keep the list so as not to evaluate the lazy values if there is no need. Thus in the case of t-raw the pieces do not need to be joined and this is the most frequent case for the value "0".

Reply:
Then the default value for the second case should be a literal. And it should probably have a comment to explain this oddity.

This also changes the body storage from being a key to which template designers don't have access (the *number* `0`) to one which they do (the *string* 0). Which can lead to odd / unfortunate situations e.g.
```XML
                <t t-set="0" t-value="[1, 2, 3]"/>
                <t t-esc="0"/>
```
will break during rendering, because the list will pass through and explode during the final join.

Reply:
I am going to put a comment concerning this value and the fact that it only unwinds when it is used.

`
                <t t-set="0" t-value="[1, 2, 3]"/>
                <t t-esc="0"/>
`
this case is not supported. The value 0 is specific for the t-call and the t-set has no t-value, only set by content. I'll add a raise to be sure..